x/posts/keeper/polls.go

```diff
@@ -113,9 +135,15 @@ func (k Keeper) GetUserAnswer(ctx sdk.Context, subspaceID uint64, postID uint64,
 }
 
 // DeleteUserAnswer deletes the user answer from the provided poll
-func (k Keeper) DeleteUserAnswer(ctx sdk.Context, subspaceID uint64, postID uint64, pollID uint32, user string) {
+func (k Keeper) DeleteUserAnswer(ctx sdk.Context, subspaceID uint64, postID uint64, pollID uint32, user string, isTally bool) {
 	store := ctx.KVStore(k.storeKey)
+	if !k.HasUserAnswer(ctx, subspaceID, postID, pollID, user) {
+		return
+	}
+
 	store.Delete(types.PollAnswerStoreKey(subspaceID, postID, pollID, user))
 
-	k.AfterPollAnswerDeleted(ctx, subspaceID, postID, pollID, user)
+	if !isTally {
+		k.AfterPollAnswerDeleted(ctx, subspaceID, postID, pollID, user)
+	}
```

Comment:
AfterPollAnswerDeleted should not be called if the deletion is from tally operation.

Reply:
Why was this decided? I believe we can call it safely in every occasion: if we are removing a user answer from the store, we should notify it anyway. It will be the listener's job to decide whether to handle it or not. And since the tally emits a custom event, it should be easy enough to distinguish the two cases if the listener wants

Reply:
Why was this decided? I believe we can call it safely in every occasion: if we are removing a user answer from the store, we should notify it anyway. It will be the listener's job to decide whether to handle it or not. And since the tally emits a custom event, it should be easy enough to distinguish the two cases if the listener wants

Reply:
@dadamu I only have this comment left. Once this is solved, we can merge this

Reply:
From my side, we can avoid from calling redundant `AfterPollAnswerDeleted` hooks since `AfterPollVotingPeriodEnded` handles the everything.

For example, when calling `AfterPollAnswerDeleted`, the listener can remove a user answer, like `DELETE FROM user_answer WHERE user_address = "cosmos...."`. In the other hand, `AfterPollVotingPeriodEnded` can remove all the answers with `DELETE FROM user_answer WHERE poll_id = 1`.

Additionally, it is better performance if the listener module wants to handle the tallied poll only and no executions with answers.

Reply:
The listener module handling the tallied poll only and no execution with answers in the current way will be like:
```go
func (h Hooks) AfterPollAnswerDeleted(sdk.Context, subspaceID uint64, postID uint64, pollID uint32, user string) {
    if !IsTalliedPoll(ctx, subspaceID, postID, pollID) {
        handleAnswer(ctx, subspaceID, postID, pollID)
    }
}
```
We need to check the poll is tallied or not in every answer deleted action, which costs O(N) redundantly where N = the amount of answers in the poll.

Reply:
@RiccardoM `Tally` function should not call `AfterPollAnswerDeleted` from my side. The main thing is to avoid passing `isTally` to `DeleteUserAnswer`, right?
If so, I think the best way is to have a new function `TallyUserAnswer`, instead of calling `DeleteUserAnswer`.

Reply:
@dadamu I'm honestly not convinced of this change to remove the call of `AfterPollAnswerDeleted` from `DeleteUserAnswer`. I think it's not logically correct: we are removing a user answer, so why not emit the event? Also, I think this is out of scope for this PR. I would like to ask you to revert this change and open another PR where we can have a discussion about this without blocking this one (that should only be related to tests and not breaking changes - like this one)

Reply:
Agree with you, open another PR. Reveted.